protocol/messenger_status_updates.go

```diff
@@ -0,0 +1,247 @@
+	shouldBroadcastUserStatus, err := m.settings.ShouldBroadcastUserStatus()
+	if err != nil {
+		return err
+	}
+
+	if !shouldBroadcastUserStatus {
+		m.logger.Debug("user status should not be broadcasted")
+		return nil
+	}
+
+	status, err := m.GetCurrentUserStatus()
+	if err != nil {
+		m.logger.Debug("Error obtaining latest status", zap.Error(err))
+		return err
+	}
+
+	status.Clock = uint64(time.Now().Unix())
+
+	err = m.settings.SaveSetting("current-user-status", status)
+	if err != nil {
+		return err
+	}
+
+	statusUpdate := &protobuf.StatusUpdate{
+		Clock:      status.Clock,
+		StatusType: protobuf.StatusUpdate_StatusType(status.StatusType),
+		CustomText: status.CustomText,
+	}
+
+	encodedMessage, err := proto.Marshal(statusUpdate)
+	if err != nil {
+		return err
+	}
+
+	rawMessage := common.RawMessage{
+		LocalChatID:         community.StatusUpdatesChannelID(),
+		Payload:             encodedMessage,
+		MessageType:         protobuf.ApplicationMetadataMessage_STATUS_UPDATE,
+		ResendAutomatically: true,
+	}
+
+	_, err = m.sender.SendPublic(context.Background(), rawMessage.LocalChatID, rawMessage)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (m *Messenger) broadcastLatestUserStatus() {
+	m.logger.Debug("broadcasting user status")
+	m.sendCurrentUserStatus()
+	go func() {
+		for {
+			select {
+			case <-time.After(5 * time.Minute):
+				m.sendCurrentUserStatus()
+			case <-m.quit:
+				return
+			}
+		}
+	}()
+}
+
+func (m *Messenger) SetUserStatus(newStatus int, newCustomText string) error {
+	if len([]rune(newCustomText)) > maxStatusMessageText {
+		return fmt.Errorf("custom text shouldn't be longer than %d", maxStatusMessageText)
+	}
+
+	if newStatus != int(protobuf.StatusUpdate_ONLINE) && newStatus != int(protobuf.StatusUpdate_DO_NOT_DISTURB) {
+		return fmt.Errorf("unknown status type")
+	}
+
+	currStatus, err := m.GetCurrentUserStatus()
+	if err != nil {
+		m.logger.Debug("Error obtaining latest status", zap.Error(err))
+		return err
+	}
+
+	if newStatus == currStatus.StatusType && newCustomText == currStatus.CustomText {
+		m.logger.Debug("Status type did not change")
+		return nil
+	}
+
+	currStatus.StatusType = newStatus
+	currStatus.CustomText = newCustomText
+
+	return m.sendUserStatus(*currStatus)
+}
+
+func (m *Messenger) HandleStatusUpdate(state *ReceivedMessageState, statusMessage protobuf.StatusUpdate) error {
+	if err := ValidateStatusUpdate(statusMessage); err != nil {
+		return err
+	}
+
+	currentStatus, err := m.GetCurrentUserStatus()
+	if err != nil {
+		m.logger.Debug("Error obtaining latest status", zap.Error(err))
+		return err
+	}
+
+	if common.IsPubKeyEqual(state.CurrentMessageState.PublicKey, &m.identity.PublicKey) { // Status message is ours
+		if currentStatus.Clock >= statusMessage.Clock || (currentStatus.StatusType == int(statusMessage.StatusType) && currentStatus.CustomText == statusMessage.CustomText) {
```

Comment:
This does not look exactly correct, I think there might be instances where it might fail.

Thinking of a case like:

Say we have 2 paired devices:

1) Device 1 sets status  `S1` (assume type is always the same) to `Clock: 0, CustomText: "A"`
2) Device 2 receives the update, goes offline
3) Device 1 sets status `S2` to `Clock: 1, customText: "B"`
4) Device 1 sets status `S3` to `Clock: 2, customText: "A"`
5) Device 2 goes back online and receive first  `S3`, which will be ignored as per code above. Then receives `S2` and it will be set to `S2`, but it should be `S3` (similarly if a message is dropped)

I think we should only trust `Clock`, and discard if clock is less (or equal, up to us), as it's effectively an updated status, as we can't guarantee that there were changes in between.